File: tests/test_exceptions/test_validation_error_equality.js
const { ValidationError } = require('../../django/core/exceptions');

describe('ValidationError Equality Tests', () => {
    test('ValidationErrors with identical string messages should be equal', () => {
        const error1 = new ValidationError("This field is required.");
        const error2 = new ValidationError("This field is required.");
        expect(error1).toEqual(error2);
    });

    test('ValidationErrors with same messages in different order should be equal', () => {
        const error3 = new ValidationError(["Error A", "Error B"]);
        const error4 = new ValidationError(["Error B", "Error A"]);
        expect(error3).toEqual(error4);
    });

    test('ValidationErrors with same field errors in different order should be equal', () => {
        const error5 = new ValidationError({"field1": ["Error 1"], "field2": ["Error 2"]});
        const error6 = new ValidationError({"field2": ["Error 2"], "field1": ["Error 1"]});
        expect(error5).toEqual(error6);
    });

    test('ValidationErrors with different messages should not be equal', () => {
        const error7 = new ValidationError("Different message");
        const error8 = new ValidationError("Another message");
        expect(error7).not.toEqual(error8);
    });

    test('ValidationError should not equal non-ValidationError objects', () => {
        const error = new ValidationError("Test message");
        expect(error).not.toEqual("Test message");
        expect(error).not.toEqual({});
        expect(error).not.toEqual(null);
    });

    test('ValidationErrors with different field sets should not be equal', () => {
        const error1 = new ValidationError({"field1": ["Error 1"]});
        const error2 = new ValidationError({"field2": ["Error 1"]});
        expect(error1).not.toEqual(error2);
    });

    test('ValidationError with error_dict should not equal one with error_list', () => {
        const error1 = new ValidationError({"field1": ["Error 1"]});
        const error2 = new ValidationError(["Error 1"]);
        expect(error1).not.toEqual(error2);
    });

    test('Empty ValidationErrors should be equal', () => {
        const error1 = new ValidationError([]);
        const error2 = new ValidationError([]);
        expect(error1).toEqual(error2);
    });

    test('ValidationErrors with nested ValidationError objects should be equal', () => {
        const nestedError1 = new ValidationError("Nested error");
        const nestedError2 = new ValidationError("Nested error");
        const error1 = new ValidationError([nestedError1]);
        const error2 = new ValidationError([nestedError2]);
        expect(error1).toEqual(error2);
    });
});

File: test_validation_error_equality.py
import pytest
from django.core.exceptions import ValidationError


def test_issue_reproduction():
    # Test basic equality - identical ValidationErrors should be equal
    error1 = ValidationError('This field is required.')
    error2 = ValidationError('This field is required.')
    assert error1 == error2, "ValidationErrors with identical messages should be equal"
    
    # Test equality with code parameter
    error3 = ValidationError('Invalid value', code='invalid')
    error4 = ValidationError('Invalid value', code='invalid')
    assert error3 == error4, "ValidationErrors with identical messages and codes should be equal"
    
    # Test equality with params
    error5 = ValidationError('Value %(value)s is invalid', params={'value': 'test'})
    error6 = ValidationError('Value %(value)s is invalid', params={'value': 'test'})
    assert error5 == error6, "ValidationErrors with identical messages and params should be equal"
    
    # Test inequality with different messages
    error7 = ValidationError('This field is required.')
    error8 = ValidationError('This field is invalid.')
    assert error7 != error8, "ValidationErrors with different messages should not be equal"
    
    # Test inequality with different codes
    error9 = ValidationError('Invalid value', code='invalid')
    error10 = ValidationError('Invalid value', code='required')
    assert error9 != error10, "ValidationErrors with different codes should not be equal"
    
    # Test equality with list of errors (order-independent)
    error11 = ValidationError(['Error 1', 'Error 2'])
    error12 = ValidationError(['Error 2', 'Error 1'])
    assert error11 == error12, "ValidationErrors with same errors in different order should be equal"
    
    # Test equality with dict of errors (order-independent)
    error13 = ValidationError({'field1': ['Error 1'], 'field2': ['Error 2']})
    error14 = ValidationError({'field2': ['Error 2'], 'field1': ['Error 1']})
    assert error13 == error14, "ValidationErrors with same field errors in different order should be equal"
    
    # Test nested ValidationErrors
    nested1 = ValidationError(ValidationError('Nested error'))
    nested2 = ValidationError(ValidationError('Nested error'))
    assert nested1 == nested2, "Nested ValidationErrors with identical content should be equal"
    
    # Test hash consistency (ValidationErrors should be hashable if they're equal)
    error15 = ValidationError('Test error')
    error16 = ValidationError('Test error')
    if error15 == error16:
        assert hash(error15) == hash(error16), "Equal ValidationErrors should have equal hashes"